src/main/java/teammates/storage/search/SearchManager.java

```diff
@@ -69,21 +72,32 @@ SolrQuery getBasicQuery(String queryString) {
         return query;
     }
 
-    QueryResponse performQuery(SolrQuery query) throws SearchNotImplementedException {
+    QueryResponse performQuery(SolrQuery query) throws SearchServiceException {
         if (client == null) {
-            throw new SearchNotImplementedException();
+            throw new SearchServiceException("Search service is not implemented.", HttpStatus.SC_NOT_IMPLEMENTED);
         }
 
         QueryResponse response = null;
 
         try {
             response = client.query(getCollectionName(), query);
         } catch (SolrServerException e) {
-            log.severe(String.format(ERROR_SEARCH_DOCUMENT, query.getQuery(), e.getRootCause())
+            Throwable rootCause = e.getRootCause();
+            log.severe(String.format(ERROR_SEARCH_DOCUMENT, query.getQuery(), rootCause)
                     + TeammatesException.toStringWithStackTrace(e));
+            if (rootCause instanceof ConnectException) {
+                throw new SearchServiceException("Error connecting to search service.",
+                        e, HttpStatus.SC_BAD_GATEWAY);
+            } else if (rootCause instanceof SocketTimeoutException) {
+                throw new SearchServiceException("A timeout was reached while processing the request.",
+                        e, HttpStatus.SC_GATEWAY_TIMEOUT);
+            } else {
+                throw new SearchServiceException("An error has occurred.", e, HttpStatus.SC_BAD_GATEWAY);
```

Comment:
This is very vague when displayed on the front-end. At the very least, should mention "An error has occurred *while performing search*".